File: skills/math-expression-parser/reference/src/evaluate.test.ts
import { describe, test, expect } from "bun:test";
import { calc } from "./evaluate";

describe("calc (end-to-end)", () => {
  describe("basic arithmetic", () => {
    test.each([
      { expr: "1 + 2", expected: 3 },
      { expr: "10 - 3", expected: 7 },
      { expr: "4 * 5", expected: 20 },
      { expr: "15 / 4", expected: 3.75 },
      { expr: "10 % 3", expected: 1 },
      { expr: "2 ** 8", expected: 256 },
    ])("$expr = $expected", ({ expr, expected }) => {
      expect(calc(expr)).toBe(expected);
    });
  });

  describe("precedence", () => {
    test.each([
      { expr: "2 + 3 * 4", expected: 14 },
      { expr: "2 * 3 + 4", expected: 10 },
      { expr: "10 - 2 * 3", expected: 4 },
      { expr: "2 + 3 ** 2", expected: 11 },
      { expr: "2 * 3 ** 2", expected: 18 },
      { expr: "2 ** 3 * 4", expected: 32 },
    ])("$expr = $expected", ({ expr, expected }) => {
      expect(calc(expr)).toBe(expected);
    });
  });

  describe("parentheses", () => {
    test.each([
      { expr: "(2 + 3) * 4", expected: 20 },
      { expr: "2 * (3 + 4)", expected: 14 },
      { expr: "(2 + 3) * (4 + 5)", expected: 45 },
      { expr: "((1 + 2) * (3 + 4))", expected: 21 },
      { expr: "(10)", expected: 10 },
    ])("$expr = $expected", ({ expr, expected }) => {
      expect(calc(expr)).toBe(expected);
    });
  });

  describe("associativity", () => {
    test.each([
      { expr: "1 - 2 - 3", expected: -4 },
      { expr: "1 - 2 + 3", expected: 2 },
      { expr: "12 / 3 / 2", expected: 2 },
      { expr: "2 ** 3 ** 2", expected: 512 },
    ])("$expr = $expected", ({ expr, expected }) => {
      expect(calc(expr)).toBe(expected);
    });
  });

  describe("unary minus", () => {
    test.each([
      { expr: "-5", expected: -5 },
      { expr: "--5", expected: 5 },
      { expr: "-(-5)", expected: 5 },
      { expr: "2 * -3", expected: -6 },
      { expr: "-2 ** 2", expected: 4 },
      { expr: "-(2 ** 2)", expected: -4 },
    ])("$expr = $expected", ({ expr, expected }) => {
      expect(calc(expr)).toBe(expected);
    });
  });

  describe("decimals", () => {
    test.each([
      { expr: "0.1 + 0.2", expected: 0.1 + 0.2 },
      { expr: "3.14 * 2", expected: 6.28 },
      { expr: ".5 + .5", expected: 1 },
    ])("$expr = $expected", ({ expr, expected }) => {
      expect(calc(expr)).toBe(expected);
    });
  });

  describe("complex expressions", () => {
    test.each([
      { expr: "2 + 3 * 4 - 1", expected: 13 },
      { expr: "(2 + 3) * (4 - 1) / 5", expected: 3 },
      { expr: "10 % 3 + 2 ** 3", expected: 9 },
      { expr: "2 ** (1 + 2)", expected: 8 },
      { expr: "100 / 10 / 2 + 3", expected: 8 },
    ])("$expr = $expected", ({ expr, expected }) => {
      expect(calc(expr)).toBe(expected);
    });
  });

  describe("errors", () => {
    test("empty expression", () => {
      expect(() => calc("")).toThrow("Empty expression");
    });

    test("whitespace only", () => {
      expect(() => calc("   ")).toThrow("Empty expression");
    });

    test("division by zero", () => {
      expect(() => calc("1 / 0")).toThrow("Division by zero");
    });

    test("modulo by zero", () => {
      expect(() => calc("5 % 0")).toThrow("Modulo by zero");
    });

    test("unmatched paren", () => {
      expect(() => calc("(2 + 3")).toThrow();
    });

    test("invalid character", () => {
      expect(() => calc("2 @ 3")).toThrow();
    });

    test("trailing operator", () => {
      expect(() => calc("2 +")).toThrow();
    });
  });
});

File: research/experiments/sweep/runs/mathexpr-full-kotlin-sonnet45/src/test/kotlin/mathexpr/EvaluateTest.kt
package mathexpr

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith

class EvaluateTest {
    @Test
    fun `evaluates simple addition`() {
        assertEquals(3.0, calc("1 + 2"))
    }

    @Test
    fun `evaluates simple subtraction`() {
        assertEquals(7.0, calc("10 - 3"))
    }

    @Test
    fun `evaluates simple multiplication`() {
        assertEquals(20.0, calc("4 * 5"))
    }

    @Test
    fun `evaluates simple division`() {
        assertEquals(3.75, calc("15 / 4"))
    }

    @Test
    fun `evaluates simple modulo`() {
        assertEquals(1.0, calc("10 % 3"))
    }

    @Test
    fun `evaluates simple power`() {
        assertEquals(256.0, calc("2 ** 8"))
    }

    @Test
    fun `evaluates addition and multiplication precedence`() {
        assertEquals(14.0, calc("2 + 3 * 4"))
    }

    @Test
    fun `evaluates multiplication and addition precedence`() {
        assertEquals(10.0, calc("2 * 3 + 4"))
    }

    @Test
    fun `evaluates subtraction and multiplication precedence`() {
        assertEquals(4.0, calc("10 - 2 * 3"))
    }

    @Test
    fun `evaluates addition and power precedence`() {
        assertEquals(11.0, calc("2 + 3 ** 2"))
    }

    @Test
    fun `evaluates multiplication and power precedence`() {
        assertEquals(18.0, calc("2 * 3 ** 2"))
    }

    @Test
    fun `evaluates power and multiplication precedence`() {
        assertEquals(32.0, calc("2 ** 3 * 4"))
    }

    @Test
    fun `evaluates parentheses changing precedence`() {
        assertEquals(20.0, calc("(2 + 3) * 4"))
    }

    @Test
    fun `evaluates parentheses with multiplication`() {
        assertEquals(14.0, calc("2 * (3 + 4)"))
    }

    @Test
    fun `evaluates multiple parentheses groups`() {
        assertEquals(45.0, calc("(2 + 3) * (4 + 5)"))
    }

    @Test
    fun `evaluates nested parentheses`() {
        assertEquals(21.0, calc("((1 + 2) * (3 + 4))"))
    }

    @Test
    fun `evaluates single parenthesized number`() {
        assertEquals(10.0, calc("(10)"))
    }

    @Test
    fun `evaluates left associative subtraction`() {
        assertEquals(-4.0, calc("1 - 2 - 3"))
    }

    @Test
    fun `evaluates left associative mixed addition and subtraction`() {
        assertEquals(2.0, calc("1 - 2 + 3"))
    }

    @Test
    fun `evaluates left associative division`() {
        assertEquals(2.0, calc("12 / 3 / 2"))
    }

    @Test
    fun `evaluates right associative power`() {
        assertEquals(512.0, calc("2 ** 3 ** 2"))
    }

    @Test
    fun `evaluates unary minus`() {
        assertEquals(-5.0, calc("-5"))
    }

    @Test
    fun `evaluates double unary minus`() {
        assertEquals(5.0, calc("--5"))
    }

    @Test
    fun `evaluates parenthesized unary minus`() {
        assertEquals(5.0, calc("-(-5)"))
    }

    @Test
    fun `evaluates multiplication with unary minus`() {
        assertEquals(-6.0, calc("2 * -3"))
    }

    @Test
    fun `evaluates unary minus with power - case 1`() {
        assertEquals(4.0, calc("-2 ** 2"))
    }

    @Test
    fun `evaluates unary minus with power - case 2`() {
        assertEquals(-4.0, calc("-(2 ** 2)"))
    }

    @Test
    fun `evaluates decimal multiplication`() {
        assertEquals(6.28, calc("3.14 * 2"))
    }

    @Test
    fun `evaluates decimal addition`() {
        assertEquals(1.0, calc(".5 + .5"))
    }

    @Test
    fun `evaluates complex expression 1`() {
        assertEquals(13.0, calc("2 + 3 * 4 - 1"))
    }

    @Test
    fun `evaluates complex expression 2`() {
        assertEquals(3.0, calc("(2 + 3) * (4 - 1) / 5"))
    }

    @Test
    fun `evaluates complex expression 3`() {
        assertEquals(9.0, calc("10 % 3 + 2 ** 3"))
    }

    @Test
    fun `evaluates power with parenthesized exponent`() {
        assertEquals(8.0, calc("2 ** (1 + 2)"))
    }

    @Test
    fun `evaluates complex expression 4`() {
        assertEquals(8.0, calc("100 / 10 / 2 + 3"))
    }

    @Test
    fun `throws error on empty string`() {
        val ex = assertFailsWith<IllegalArgumentException> {
            calc("")
        }
        assert(ex.message!!.contains("Empty expression"))
    }

    @Test
    fun `throws error on whitespace only`() {
        val ex = assertFailsWith<IllegalArgumentException> {
            calc("   ")
        }
        assert(ex.message!!.contains("Empty expression"))
    }

    @Test
    fun `throws error on division by zero`() {
        val ex = assertFailsWith<ArithmeticException> {
            calc("1 / 0")
        }
        assert(ex.message!!.contains("Division by zero"))
    }

    @Test
    fun `throws error on modulo by zero`() {
        val ex = assertFailsWith<ArithmeticException> {
            calc("5 % 0")
        }
        assert(ex.message!!.contains("Modulo by zero"))
    }

    @Test
    fun `throws error on missing closing paren`() {
        val ex = assertFailsWith<IllegalArgumentException> {
            calc("(2 + 3")
        }
        assert(ex.message!!.contains("rparen"))
    }

    @Test
    fun `throws error on unexpected character`() {
        val ex = assertFailsWith<IllegalArgumentException> {
            calc("2 @ 3")
        }
        assert(ex.message!!.contains("Unexpected character"))
    }

    @Test
    fun `throws error on incomplete expression`() {
        val ex = assertFailsWith<IllegalArgumentException> {
            calc("2 +")
        }
        assert(ex.message!!.contains("Unexpected end of input"))
    }
}
